Backend + Feature: Add shuriken highlight

SuperCl4sh opened the pull request:
## What
Highlight mobs if they've been hit with an Extremely Real Shuriken. Also fixes Mob.HurtEvent not updating nametags and adds shurikenUsed to Mob.kt.

## Changelog New Features
+ Highlight mobs if they've been hit with an Extremely Real Shuriken. - SuperClash

## Changelog Technical Details
+ Added shurikenUsed to Mob.kt. - SuperClash
+ Fixed Mob class not properly updating health of mobs when hurt. - Thunderblade73
    * Fixed Armorstand being retained by our Internal Mob Structure, now they get updated with the respawning of the Armorstand.

SuperCl4sh committed:
Add shuriken highlight

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/data/mob/Mob.kt:
```diff
@@ -115,6 +118,7 @@ class Mob(
 
     val isCorrupted get() = !RiftApi.inRift() && baseEntity.isCorrupted() // Can change
     val isRunic = !RiftApi.inRift() && baseEntity.isRunic() // Does not Change
+    val shurikenUsed get() = armorStand?.getNameAsString()?.endsWith("§b✯") == true && mobType != Type.DUNGEON
```
Flip the check. As the second one is *free* but the first one isn't.

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ShurikenHighlight.kt:
```diff
@@ -0,0 +1,31 @@
+object ShurikenHighlight {
+    private val config get() = SkyHanniMod.feature.combat.mobs
+
+    @HandleEvent(onlyOnSkyblock = true)
```
onlyOnSkyblock is redundant as the event only fires when ``onlyOnSkyblock == true``

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ShurikenHighlight.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    @HandleEvent(onlyOnSkyblock = true)
```
onlyOnSkyblock is redundant as the event only fires when ``onlyOnSkyblock == true``

jani270 reviewed (dismissed):
Please move the `Added shurikenUsed to Mob.kt - SuperClash` Changelog to the Technical Details Changelog, as this change is not visible to the user.

Thunderblade73 commented:
> Please move the `Added shurikenUsed to Mob.kt - SuperClash` Changelog to the Technical Details Changelog, as this change is not visible to the user.

It is visible in the fact that the mod runs smoother (as we do not stack up armorstand). The changelog is simply not accurate, but I wrote a better version of it already, which is also now set.

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ShurikenHighlight.kt:
```diff
@@ -0,0 +1,31 @@
+        if (!config.shurikenHighlight) return
+        DelayedRun.runNextTick { // mob.shurikenUsed isn't updated yet, so wait a tick before checking
+            if (event.mob.shurikenUsed) {
+                event.mob.highlight(config.shurikenColour.toSpecialColor())
```
Add to the condition "config.shurikenHighlight" so that it will work properly when disabling the feature.

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ShurikenHighlight.kt:
```diff
@@ -0,0 +1,31 @@
+    fun onMobHurt(event: MobEvent.Hurt.SkyblockMob) {
+        if (!config.shurikenHighlight) return
+        DelayedRun.runNextTick { // mob.shurikenUsed isn't updated yet, so wait a tick before checking
+            if (config.shurikenHighlight && event.mob.shurikenUsed) {
+                event.mob.highlight(config.shurikenHighlightColour.toSpecialColor())
+            }
```
Not what I meant.
```suggestion
            if (event.mob.shurikenUsed) {
                event.mob.highlight(config.shurikenHighlightColour.toSpecialColor()) { config.shurikenHighlight }
            }
```

SuperCl4sh replied:
Forgot that was a thing oops.

SuperCl4sh committed:
Update src/main/java/at/hannibal2/skyhanni/features/combat/mobs/ShurikenHighlight.kt

Co-authored-by: Thunderblade73 <85900443+Thunderblade73@users.noreply.github.com>

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/add-shuriken

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/config/features/combat/MobsConfig.kt
#	src/main/java/at/hannibal2/skyhanni/data/mob/Mob.kt

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/add-shuriken

CalMWolfs committed:
fix merge

hannibal002 commented:
Fixed Armorstand being retained by our Internal Mob Structure. - Thunderblade73 

please change the wording here, retained and mob structure is not relevant in a user facing changelog

SuperCl4sh commented:
Does 
+ Fixed Mob class not properly updating health of mobs when hurt. - Thunderblade73
    * Fixed Armorstand being retained by our Internal Mob Structure, now they get updated with the respawning of the Armorstand.
   
work?

Thunderblade73 commented:
> Does
> 
> * Fixed Mob class not properly updating health of mobs when hurt. - Thunderblade73
>   
>   * Fixed Armorstand being retained by our Internal Mob Structure, now they get updated with the respawning of the Armorstand.
> 
> work?

yes.

hannibal002 reviewed (changes requested):
> Fixed Mob class not properly updating health of mobs when hurt. - Thunderblade73 

what does this mean for the users?
users might ask: what is a "mob class"? (users dont need to know that)
isntead, explain what impact this has on users. what features will work now correctly, when would those features not work. what has chagned on a "user facing" side.
everything else belongs in the technical changelog explanation if standalone or just in the norrmal pr description.

i dont wanna be picky, i just dont wanna have technical text in the user facing chagnelog that avoids the users needing to ask "what does this mean"

Thunderblade73 commented:
> > Fixed Mob class not properly updating health of mobs when hurt. - Thunderblade73
> 
> what does this mean for the users? users might ask: what is a "mob class"? (users dont need to know that) isntead, explain what impact this has on users. what features will work now correctly, when would those features not work. what has chagned on a "user facing" side. everything else belongs in the technical changelog explanation if standalone or just in the norrmal pr description.
> 
> i dont wanna be picky, i just dont wanna have technical text in the user facing chagnelog that avoids the users needing to ask "what does this mean"

Yeah it should be a Technical change. (I did not catch that it was moved to fixes).
Since yes, there is a user facing improvement with a minimal performance improvement (as we do retrain not invalid objects anymore) but for most/all player it won't be noticeable.

hannibal002 commented:
@SuperCl4sh bump

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/SuperCl4sh/add-shuriken

SuperCl4sh commented:
Oops, I forgot. I think it should be good now:tm:,

SuperCl4sh commented:
Going to move this into damage indicator so it works better with twilight highlight (and other features in the future).
